lib/integrations/awsra/createsession/createsession.go

```diff
@@ -0,0 +1,301 @@
+
+	// Task 1-6
+	canonicalRequestBodyHash := sha256.Sum256(canonicalRequestBody)
+
+	// Task 1-7
+	canonicalRequestString := fmt.Sprintf(`POST
+/sessions
+
+content-type:application/json
+host:%s
+x-amz-date:%s
+x-amz-x509:%s
+
+%s
+%x`,
+		canonicalRequest.Header.Get("Host"),
+		canonicalRequest.Header.Get("X-Amz-Date"),
+		canonicalRequest.Header.Get("X-Amz-X509"),
+		signedHeaders,
+		canonicalRequestBodyHash,
+	)
+
+	// Task 1-8
+	canonicalRequestHashBytes := sha256.New()
+	canonicalRequestHashBytes.Write([]byte(canonicalRequestString))
+	canonicalRequestHash := hex.EncodeToString(canonicalRequestHashBytes.Sum(nil))
+
+	// Task 2: Create a string to sign
+	// https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-sign-process.html#authentication-task2
+
+	// Teleport only uses ECDSA keys in Roles Anywhere integration.
+	const algorithm = "AWS4-X509-ECDSA-SHA256"
+	credentialScope := formatedDate[:8] + "/" + req.awsRegion + "/rolesanywhere/aws4_request"
+
+	stringToSign := algorithm + "\n" +
+		formatedDate + "\n" +
+		credentialScope + "\n" +
+		canonicalRequestHash
+
+	// Task 3: Calculate the signature
+	// https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-sign-process.html#authentication-task3
+	signatureHash := sha256.Sum256([]byte(stringToSign))
+	signature, err := ecdsa.SignASN1(rand.Reader, req.PrivateKey, signatureHash[:])
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	// Task 4-2
+	credentialString := req.Certificate.SerialNumber.String() + "/" + credentialScope
+	// Task 4-3
+	canonicalRequest.Header.Set("Authorization",
+		algorithm+" "+
+			"Credential="+credentialString+", "+
+			"SignedHeaders="+signedHeaders+", "+
+			"Signature="+hex.EncodeToString(signature),
+	)
+
+	resp, err := req.httpClient.Do(canonicalRequest)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	defer resp.Body.Close()
+
+	respBody, err := utils.ReadAtMost(resp.Body, teleport.MaxHTTPResponseSize)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	if resp.StatusCode != http.StatusCreated {
+		return nil, trace.ReadError(resp.StatusCode, respBody)
+	}
+
+	// https://docs.aws.amazon.com/rolesanywhere/latest/userguide/authentication-create-session.html#response-syntax
+	createSessionResp := struct {
+		CredentialSet []struct {
+			Credentials struct {
+				AccessKeyId     string `json:"accessKeyId"`
+				SecretAccessKey string `json:"secretAccessKey"`
+				SessionToken    string `json:"sessionToken"`
+				Expiration      string `json:"expiration"`
+			} `json:"credentials"`
+		} `json:"credentialSet"`
+	}{}
+
+	if err := json.Unmarshal(respBody, &createSessionResp); err != nil {
+		return nil, trace.BadParameter("parsing response: %v", err)
+	}
+
+	if len(createSessionResp.CredentialSet) == 0 {
+		return nil, trace.BadParameter("no credentials received from ")
+	}
+
+	credentials := createSessionResp.CredentialSet[0].Credentials
```

Comment:
Is it possible to receive multiple credentials? If this happens, is it ok to silently drop the additional creds? Should we error? warn about the unexpected response? conitnue?